File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Model/Statement/IfStatement.java
package Model.Statement;

import Model.Expression.ExpressionException;
import Model.Expression.IExpression;
import Model.State.ProgramState;
import Util.Stack.MyIStack;

public class IfStatement implements IStatement {
    private IExpression expression;
    private IStatement thenStatement;
    private IStatement elseStatement;

    public IfStatement(IExpression expression, IStatement thenS, IStatement elseS){
        this.expression = expression;
        this.thenStatement = thenS;
        this.elseStatement = elseS;
    }

    @Override
    public ProgramState execute(ProgramState programState) throws ExpressionException {
        int result = this.expression.evaluate(programState.getSymbolTable(),programState.getHeap());

        MyIStack<IStatement> executionStack = programState.getExecutionStack();

        if( result != 0) { // true
            executionStack.push(this.thenStatement);
        }
        else
            executionStack.push(this.elseStatement); // false

        return programState;
    }

    @Override
    public String toString() {
        return "(" + "IF " + expression + " THEN " + thenStatement + " ELSE " + elseStatement + ")";
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Model/Statement/IfStatement.cs
using ToyLanguage.Model.Expression;
using ToyLanguage.Model.State;

namespace ToyLanguage.Model.Statement
{
    class IfStatement : IStatement
    {
        private readonly IExpression condition;
        private readonly IStatement thenStatement;
        private readonly IStatement elseStatement;

        public IfStatement(IExpression condition, IStatement thenStatement, IStatement elseStatement)
        {
            this.condition = condition;
            this.thenStatement = thenStatement;
            this.elseStatement = elseStatement;
        }

        public ProgramState Execute(ProgramState programState)
        {
            int value = condition.Evaluate(programState.SymbolTable);

            if( value != 0)
            {
                programState.ExecutionStack.Push(thenStatement);
            }
            else
            {
                programState.ExecutionStack.Push(elseStatement);
            }

            return programState;
        }

        public override string ToString()
        {
            return "IfStatement(" + condition + ")" + "Then " + thenStatement + " Else " + elseStatement;
        }
    }
}
